pyuvm/s18_pyuvm_reg_block.py

```diff
@@ -1,19 +1,182 @@
+        # add to the BLK main mapping
+        if (in_blk not in self.child_blk):
+            self.blk_maping[in_blk.get_name()] = 1
+            self.child_blk.append(in_blk)
+
+    # blk_add_register
     def _add_register(self, reg):
-        self._regs.append(reg)
\ No newline at end of file
+        self._regs.append(reg)
+        self.blk_set_reg_mapping(reg)
+
+    # blk_get_def_map
+    def blk_get_def_map(self):
+        return self.def_map
+
+    # get_blk_full_name
+    def get_blk_full_name(self) -> str:
+        if (self.parent_blk is None):
+            self.blk_name
+        else:
+            self.parent_blk + "." + self.blk_name
+
+    # blk_get_registers
+    def _get_registers(self) -> list:
+        local_reg_collector = []
+        if self.is_locked() is True:
+            for r in self._regs:
+                if self.blk_is_reg_mapped(r) is True:
+                    local_reg_collector.append(r)
+            if len(self.child_blk) != 0:
+                for b in self.child_blk:
+                    local_reg_collector.append(b.blk_get_registers())
+        else:
+            uvm_fatal(self.gen_message("_get_registers -- register block must be locked"))
+        return local_reg_collector
+
+    # blk_get_fields
+    def blk_get_fields(self) -> list:
+        local_field_collector = []
+        for r in self.blk_get_registers():
+            local_field_collector.append(r.get_fields())
+        return local_field_collector
+
+    # get_all_child_blk
+    def get_all_child_blk(self) -> list:
+        local_blk_collector = []
+        for b in self.child_blk:
+            if (self.blk_is_child_mapped(b) is True):
+                local_blk_collector.append(b)
+                local_blk_collector.append(b.get_all_child_blk())
+        return local_blk_collector
+
+    # blk_add_map
+    def blk_add_map(self, map_i: uvm_reg_map):
+        if (self.is_locked() is True):
+            uvm_fatal(self.gen_message("blk_add_map -- register block should be locked"))
+
+        if (self.map_mapping[map_i] is True):
+            uvm_fatal(self.header)
+        else:
+            self.maps.append(map_i)
+
+        if (self.def_map is None):
+            self.def_map = map_i
+
+    # blk_create_map byte_addressing and byte_en
+    # along with endianess not yet supported
+    def blk_create_map(self, name: str, base_addr: int):
+        lmap = uvm_reg_map.create(name, self)
+        lmap.configure(self, base_addr)
+        self.blk_add_map(lmap)
+
+    # reset_blk
+    def reset_blk(self):
+        uvm_not_implemeneted(self.gen_message("reset_blk -- not implemented"))
+
+    '''
+        TODO: the following must be completed
+        1. add_vreg
+        2. remove_reg
+        3. add_mem
+        4. remove_mem
+        5. unregister_blk
+        6. write_reg or write_mem by name
+        7. read_reg or read_mem by name
+        8. get_virtual_fields in a unique field list
+        9. get_mem to return a list of memories
+    '''
```

Comment:
new class methods + TODOs as requested. Added ` uvm_not_implemeneted(self.gen_message("justify_text"))` line anytime something is missing implementation